pkg/imgpkg/registry/keychain.go

```diff
@@ -0,0 +1,217 @@
+	var lastErr error
+
+	for i := 0; i < 5; i++ {
+		auth, lastErr = doFunc()
+		if lastErr == nil {
+			return auth, nil
+		}
+
+		if strings.Contains(lastErr.Error(), errCredentialsNotFoundMessage) || strings.Contains(lastErr.Error(), errCredentialsMissingUsernameMessage) || strings.Contains(lastErr.Error(), errCredentialsMissingServerURLMessage) {
+			return auth, lastErr
+		}
+
+		time.Sleep(2 * time.Second)
+	}
+	return auth, fmt.Errorf("Retried 5 times: %s", lastErr)
+}
+
+type envKeychain struct {
+	environFunc func() []string
+
+	infos       []envKeychainInfo
+	collectErr  error
+	collected   bool
+	collectLock sync.Mutex
+}
+
+var _ regauthn.Keychain = &envKeychain{}
+
+func (k *envKeychain) Resolve(target regauthn.Resource) (regauthn.Authenticator, error) {
+	infos, err := k.collect()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, info := range infos {
+		if info.Hostname == target.RegistryStr() {
+			return regauthn.FromConfig(regauthn.AuthConfig{
+				Username:      info.Username,
+				Password:      info.Password,
+				IdentityToken: info.IdentityToken,
+				RegistryToken: info.RegistryToken,
+			}), nil
+		}
+	}
+
+	return regauthn.Anonymous, nil
+}
+
+func (k *envKeychain) collect() ([]envKeychainInfo, error) {
+	k.collectLock.Lock()
+	defer k.collectLock.Unlock()
+
+	if k.collected {
+		return append([]envKeychainInfo{}, k.infos...), nil
+	}
+	if k.collectErr != nil {
+		return nil, k.collectErr
+	}
+
+	const (
+		globalEnvironPrefix = "IMGPKG_REGISTRY_"
+		sep                 = "_"
+	)
+
+	funcsMap := map[string]func(*envKeychainInfo, string) error{
+		"HOSTNAME": func(info *envKeychainInfo, val string) error {
+			registry, err := regname.NewRegistry(val, regname.StrictValidation)
+			if err != nil {
+				return fmt.Errorf("Parsing registry hostname: %s (e.g. gcr.io, index.docker.io)", err)
+			}
+			info.Hostname = registry.RegistryStr()
+			return nil
+		},
+		"USERNAME": func(info *envKeychainInfo, val string) error {
+			info.Username = val
+			return nil
+		},
+		"PASSWORD": func(info *envKeychainInfo, val string) error {
+			info.Password = val
+			return nil
+		},
+		"IDENTITY_TOKEN": func(info *envKeychainInfo, val string) error {
+			info.IdentityToken = val
+			return nil
+		},
+		"REGISTRY_TOKEN": func(info *envKeychainInfo, val string) error {
+			info.RegistryToken = val
+			return nil
+		},
+	}
+
+	defaultInfo := envKeychainInfo{}
+	infos := map[string]envKeychainInfo{}
+
+	for _, env := range k.environFunc() {
+		pieces := strings.SplitN(env, "=", 2)
+		if len(pieces) != 2 {
+			continue
+		}
+
+		var matched bool
+
+		if strings.HasPrefix(pieces[0], globalEnvironPrefix) {
+			authOpt := strings.TrimPrefix(pieces[0], globalEnvironPrefix)
+
+			if updateFunc, ok := funcsMap[authOpt]; ok {
+				matched = true
+				err := updateFunc(&defaultInfo, pieces[1])
+				if err != nil {
+					k.collectErr = err
+					return nil, k.collectErr
+				}
+			} else {
+				splits := strings.SplitN(authOpt, "_", 2)
```

Comment:
This isn't quite right. It would work if our "types" didn't include `_` themselves, but some do. Whoops! Back to the drawing board.